pkg/mapi2capi/aws.go

```diff
@@ -0,0 +1,431 @@
+	}
+
+	if awsProviderConfig.UserDataSecret != nil && awsProviderConfig.UserDataSecret.Name != "" {
+		capiMachineSet.Spec.Template.Spec.Bootstrap = capiv1.Bootstrap{
+			DataSecretName: &awsProviderConfig.UserDataSecret.Name,
+		}
+	}
+
+	if m.Infrastructure == nil || m.Infrastructure.Status.InfrastructureName == "" {
+		errs = append(errs, errInfrastructureInfrastructureNameCannotBeNil)
+	} else {
+		capiMachineSet.Spec.Template.Spec.ClusterName = m.Infrastructure.Status.InfrastructureName
+		capiMachineSet.Spec.ClusterName = m.Infrastructure.Status.InfrastructureName
+	}
+
+	if len(errs) > 0 {
+		return capiv1.MachineSet{}, capav1.AWSMachineTemplate{}, warnings, utilerrors.NewAggregate(errs)
+	}
+
+	return capiMachineSet, capaMachineTemplate, warnings, nil
+}
+
+// ToMachineTemplateSpec implements the ProviderSpec conversion interface for the AWS provider,
+// it converts AWSProviderSpec to AWSMachineTemplateSpec.
+//
+//nolint:funlen
+func (p AWSProviderSpecAndInfra) ToMachineTemplateSpec() (capav1.AWSMachineTemplateSpec, []string, error) {
+	var (
+		errs     []error
+		warnings []string
+	)
+
+	rootVolume, nonRootVolumes := convertAWSBlockDeviceMappingSpecToCAPI(p.Spec.BlockDevices)
+
+	capiAdditionalSecurityGroups := convertAWSSecurityGroupstoCAPI(p.Spec.SecurityGroups)
+
+	capiAdditionalAWSTags := convertAWSTagsToCAPI(p.Spec.Tags)
+
+	capiIAMInstanceProfile := convertIAMInstanceProfiletoCAPI(p.Spec.IAMInstanceProfile)
+
+	capiMetadataServiceOptions := convertMetadataServiceOptionstoCAPI(p.Spec.MetadataServiceOptions)
+
+	capiSpotMarketOptions := convertAWSSpotMarketOptionsToCAPI(p.Spec.SpotMarketOptions)
+
+	capiAWSResourceReference := convertAWSResourceReferenceToCAPI(p.Spec.Subnet)
+
+	capiAWSAMIReference, err := convertAWSAMIResourceReferenceToCAPI(p.Spec.AMI)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	spec := capav1.AWSMachineTemplateSpec{
+		Template: capav1.AWSMachineTemplateResource{
+			Spec: capav1.AWSMachineSpec{
+				AMI:                      capiAWSAMIReference,
+				AdditionalSecurityGroups: capiAdditionalSecurityGroups,
+				AdditionalTags:           capiAdditionalAWSTags,
+				// CloudInit. Not defined as we use ignition in OpenShift.
+				IAMInstanceProfile: capiIAMInstanceProfile,
+				Ignition: &capav1.Ignition{
+					Version:     "3.4",                                               // Hardcoded for OpenShift.
+					StorageType: capav1.IgnitionStorageTypeOptionUnencryptedUserData, // Hardcoded for OpenShift.
+				},
+				// ImageLookupBaseOS. Not used in OpenShift.
+				// ImageLookupFormat. Not used in OpenShift.
+				// ImageLookupOrg. Not used in OpenShift.
+
+				// TODO: what to do with instanceID? in MAPI that's not in AWSMachineProviderConfig but outside of it.
+				// Is this propagated down by the CAPA controller automatically from the CAPI Machine
+
+				// InstanceID. This is dynamically populated by the controller.
```

Comment:
Is it required/will the controller attempt to create a new instance if this isn't populated?